src/test/java/com/amazonaws/encryptionsdk/model/KeyBlobTest.java

```diff
@@ -170,4 +172,76 @@ public void checkKeyLen() {
 
         assertEquals(mockDataKey_.getEncryptedDataKey().length, reconstructedKeyBlob.getEncryptedDataKeyLen());
     }
+
+    private byte[] negativeKeyProviderIdLenTestVector() {
+	// key provider id len of -1, key provider info len of 2, and key len of 3
+        return new byte[]{
+            (byte)0xff, (byte)0xff, (byte)0x01, (byte)0x00, (byte)0x02, (byte)0x02, (byte)0x03,
+	    (byte)0x00, (byte)0x03, (byte)0x04, (byte)0x05, (byte)0x06
+        };
+    }
+
+    private byte[] negativeKeyProviderInfoLenTestVector() {
+	// key provider id len of 1, key provider info len of -2, key len of 3
+        return new byte[] {
+            (byte)0x00, (byte)0x01, (byte)0x01, (byte)0xff, (byte)0xfe, (byte)0x02, (byte)0x03,
+            (byte)0x00, (byte)0x03, (byte)0x04, (byte)0x05, (byte)0x06
+        };
+    }
+
+    private byte[] negativeKeyLenTestVector() {
+	// key provider id len of 1, key provider info len of 2, key len of -3
+        return new byte[] {
+            (byte)0x00, (byte)0x01, (byte)0x01, (byte)0x00, (byte)0x00, (byte)0x02, (byte)0x03,
+            (byte)0xff, (byte)0xfd, (byte)0x04, (byte)0x05, (byte)0x06
+        };
+    }
+
+    private void assertIncomplete(final byte[] vector) {
+	assertFalse(deserialize(vector).isComplete());
+    }
+    
+    @Test
+    public void checkNegativeKeyProviderIdLen() {
+        final byte[] keyBlobBytes = createKeyBlobBytes();
+
+        // manually set the keyProviderIdLen to negative
+        final byte[] negativeKeyProviderIdLen = ByteBuffer.allocate(Short.BYTES)
+	    .putShort((short) -1).array();
+        System.arraycopy(negativeKeyProviderIdLen, 0, keyBlobBytes, 0, Short.BYTES);
+
+	// a negative field length throws a parse exception, so deserialization is incomplete
+	assertIncomplete(keyBlobBytes);
+	assertIncomplete(negativeKeyProviderIdLenTestVector());
```

Comment:
Sorry, I should have clarified. What I'm looking for is a separate test that calls the full deserialize path on a byte array crafted to contain the bad values.

That being said, the vectors you provided triggered paths in my brain that I had skipped over, that change what this issue is.

Per the message format spec[1], every one of these values is an *unsigned* integer. What would be better to be testing is to test the methods in `PrimitivesParser` to verify that none of them can result in negative values.

[1] https://docs.aws.amazon.com/encryption-sdk/latest/developer-guide/message-format.html

Reply:
I see; I will investigate `PrimitivesParser` then